app/src/main/java/com/github/h3lp3rs/h3lp/messaging/Conversation.kt

```diff
@@ -16,26 +22,153 @@ class Conversation(
     private val currentMessenger: Messenger
 ) {
     private val database = databaseOf(MESSAGES)
+    private var publicKey: PublicKey? = null
+    private val allMessages: HashMap<String, String> = HashMap()
+    private val keyAlias = "CONVERSATION_KEY_$conversationId"
+
+
+    init {
+
+        val keyStore = KeyStore.getInstance(ANDROID_KEY_STORE).apply { load(null) }
+        val alias = keyStore.aliases().toList()
+
+        if (!alias.contains(keyAlias)) {
+            val kpg =
+                KeyPairGenerator.getInstance(KeyProperties.KEY_ALGORITHM_RSA, ANDROID_KEY_STORE)
+
+            kpg.initialize(
+                KeyGenParameterSpec.Builder(
+                    keyAlias, KeyProperties.PURPOSE_ENCRYPT
+                            or KeyProperties.PURPOSE_DECRYPT
+                )
+                    .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_RSA_OAEP)
+                    .setDigests(KeyProperties.DIGEST_SHA1)
+                    .build()
+            )
+
+            val kp = kpg.generateKeyPair()
+
+            // Send public key to the database
+            val encodedPublicKey = Base64.encodeToString(kp.public.encoded, Base64.DEFAULT)
+            database.setString(
+                conversationId + "/" + KEYS_SUB_PATH + "/" + currentMessenger.name,
+                encodedPublicKey
+            )
+        }
+
+    }
 
     /**
      * Sends a message from the current user to the database
      * @param messageText The message text
      */
     fun sendMessage(messageText: String) {
-        val message = Message(currentMessenger, messageText)
-        database.addToObjectsListConcurrently(conversationId, Message::class.java, message)
+        // Retrieve public key from Bob
+        publicKey?.let { publicKey ->
+            encryptAndSend(publicKey!!, messageText)
+        }?.run {
+            val path = conversationId + "/" + KEYS_SUB_PATH + "/" +
+                    Messenger.values()[(currentMessenger.ordinal + 1) % 2].name
+
+            database.getString(path).thenApply {
+                val decodedKey = Base64.decode(it, Base64.DEFAULT)
+                val keySpecs = X509EncodedKeySpec(decodedKey)
+                publicKey = KeyFactory.getInstance("RSA").generatePublic(keySpecs)
```

Comment:
RSA should definitely be a constant if we want to change our encryption mode later on

Reply:
Oh boy if only it was that easy to change the encryption mode :)